ruby_event_store/lib/ruby_event_store/mappers/transformation/preserve_types.rb

```diff
@@ -0,0 +1,132 @@
+# frozen_string_literal: true
+
+module RubyEventStore
+  module Mappers
+    module Transformation
+      SerializationError = Class.new(ArgumentError)
+
+      class PreserveTypes
+        def initialize
+          @registered_type_serializers = {}
+        end
+
+        def register(type, serializer:, deserializer:)
+          @registered_type_serializers[type.to_s] = {
+            serializer: serializer,
+            deserializer: deserializer
+          }
+          self
+        end
+
+        def dump(record)
+          types = store_types(record.data)
+
+          Record.new(
+            event_id:   record.event_id,
+            event_type: record.event_type,
+            data:       transform_data(record.data),
+            metadata:   record.metadata.merge(types: types),
+            timestamp:  record.timestamp,
+            valid_at:   record.valid_at,
+          )
+        end
+
+        def load(record)
+          types = record.metadata.delete(:types)
+          data = types ? restore_types(record.data, types) : record.data
+
+          Record.new(
+            event_id:   record.event_id,
+            event_type: record.event_type,
+            data:       data,
+            metadata:   record.metadata,
+            timestamp:  record.timestamp,
+            valid_at:   record.valid_at,
+          )
+        end
+
+        private
+
+        SYMBOL_KEYS_KEY = '_res_symbol_keys'
+        RESERVED_KEYS = [
+          SYMBOL_KEYS_KEY, SYMBOL_KEYS_KEY.to_sym,
+        ]
+        PASS_THROUGH = ->(v) { v }
+
+        def transform_data(argument)
+          argument.each_with_object({}) do |(key, value), hash|
+            hash[key] = transform_argument(value)
+          end
+        end
+
+        def transform_argument(argument)
+          case argument
+          when Hash
+            transform_data(argument)
+          when Array
+            argument.map{|i| transform_argument(i)}
+          else
+            serializer_of(argument.class.name).call(argument)
+          end
+        end
+
+        def store_types(argument)
+          argument.each_with_object({}) do |(key, value), hash|
+            hash[serialize_hash_key(key)] = store_type(value)
+          end
+          .merge(SYMBOL_KEYS_KEY => argument.each_key.grep(Symbol).map!(&:to_s))
```

Comment:
I'd remove that logic and raise an gtfo error when someone gives us hashes with mixed strings and symbols, or at least with the same symbol defined as string (like `{ foo: 10, "foo" => 10 }`) and then just do `foo[key.to_s] || foo[key.to_sym]` when reading to try both.

Reply:
In especially that we ignore hashes with keys other than strings or symbols. So from one hand RES would be very "smart" in terms of data persistence (differentiating between symbols and strings) and on the other hand very stupid (forbidding all other kinds of hashes, i.e. hash with keys being numbers, which is perfectly valid usecase for sparse arrays). It does not click for me and is inconsistent in the approach.

Reply:
Also, the performance both in terms of CPU (data reading here is complex) and payload size, some events are big and have lot of keys, and serialization/deserialization takes time.

Reply:
Also, the serializer of the record's data (YAML, JSON, ...) does not necessarily preserve strings vs symbols differences. So:
* if we assume that strings vs. symbols are preserved, then let's make the `types` also keeping symbols and strings
* otherwise, let's just ignore it in both.

As far as I understand how it would work, we do assume that `data` could keep symbols and strings (which is not true in most cases), but it cannot keep symbols and strings in metadata (so we do the "remember the list of keys which were symbols so we can make these strings symbols again after deserialization").

Reply:
> In especially that we ignore hashes with keys other than strings or symbols.
This we could change - keys also could be transformed using provided set of serializers/deserializers - then it could be anything as long as you provide a way to dumo/load it.

Reply:
> if we assume that strings vs. symbols are preserved, then let's make the types also keeping symbols and strings

That's just another serializer for `Symbol` type.